Userland/Libraries/LibC/sys/select.h

```diff
@@ -19,6 +19,11 @@
 
 __BEGIN_DECLS
 
+// We can *not* rely on struct timeval being declared in <sys/time.h>, since
+// <sys/time.h> includes <sys/select.h>! Forward-declare it here.
+struct timeval;
+struct timespec;
```

Comment:
According to [POSIX](https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_select.h.html), `sys/select.h` _should_ provide a full definition of `timeval`, so we could just include the appropriate header from `Kernel/API/POSIX` here. `timespec` is provided through the include of `time.h` above.

Also, after the original patch that added headers, I started looking into cleaning up the existing ones that _aren't_ mandated. If I recall correctly, that included removing the `#include <sys/time.h>` in `sys/select.h`, because it didn't seem to break anything more than it already was. That would be one :yakring: taken care of.

Reply:
> `sys/select.h` _should_ provide a full definition of `timeval`

And it will, since it includes `<sys/time.h>`. The trouble is with compiling `<sys/select.h>` itself, since _at that point_ `<sys/time.h>` may not yet be fully included, if `<sys/select.h>` itself is being included from inside `<sys/time.h>`.

> so we could just include the appropriate header from `Kernel/API/POSIX` here.

Yes, we could do that alternatively.

> `timespec` is provided through the include of `time.h` above.

I wouldn't be so sure, for the same reason: what if `<time.h>` itself ends up including `<sys/select.h>`, perhaps transitively? (Maybe it doesn't today, but better be prepared for when it does.)

With these circular includes, the best way to handle this seems to be for each header to just forward-declare the structs that it needs (in addition to including the other headers that should define them).

Reply:
Personally I think including the Kernel/API headers when needed makes the most sense, as that's where the definitions are. if that makes compile times wonky though we can reconsider.

Reply:
Changed to include `<Kernel/API/POSIX/sys/time.h>` instead of `<sys/time.h>`